### PR TITLE
Structured Test Solution and Key Testing Standards

### PR DESCRIPTION
**Pull request by ziadahanass**
### What  
Added a structured document defining testing standards, guidelines, and procedures for the project.  

### Why  
To ensure consistency in how tests are written, maintained, and executed across the team.  

### How  
- Document includes testing categories (Unit, Integration, Functional, E2E).  
- Added detailed running instructions for pytest and unittest.  
- Integrated mock examples for better isolation of tests.

### CODE REVIEW

File: solutions/tests/README.md [hunk 1]
```diff
@@ -1 +1,174 @@
-# Tests
+# Structured Test Solution and Key Testing Standards
+
+## Overview
+
+This document outlines the structure, guidelines, and standards for the test
+ solutions. It serves to guide developers, quality assurance engineers, and
+  other stakeholders in understanding how to write, execute, and maintain tests
+   within the project.
```
Comment (Bikaze) [inline solutions/tests/README.md:8]: Can you tailor this to our group tests, this is all about this group and you can start this way: "This document outlines the structure, guidelines, and standards for the test written by our group for the ET code review project...."

File: solutions/tests/README.md [hunk 2]
```diff
@@ -1 +1,174 @@
...
+  - Located in the `/functional` directory.
+  - **Purpose**: Focuses on testing application functionality from the user’s perspective.
+
+- **End-to-End (E2E) Tests**:
+  - Located in the `/e2e` directory.
+  - **Purpose**: Simulates real-world user scenarios and ensures the
+   application’s workflow functions as intended.
+
+- **Fixtures**:
+  - Located in the `/fixtures` directory.
```
Comment (Bikaze) [inline solutions/tests/README.md:47]: You can possibly remove other types of tests because in this case we only care about unittests.

File: solutions/tests/README.md [hunk 3]
```diff
@@ -1 +1,174 @@
...
+
+### Writing Tests
+
+- **Isolate Tests**: Each test should be independent and should not rely on
+   shared states from other tests. This prevents flakiness.
+  
+- **Clear Assertions**: Ensure each test has clear and specific assertions to
+   check expected outcomes.
+  
+- **Mocking**: Use tools like `unittest.mock` or `pytest-mock` to mock external
```
Comment (Bikaze) [inline solutions/tests/README.md:68]: We are not supposed to use Mocking in the unittests that are required for this project.

**Review by Bikaze — COMMENTED** [review]
The amazing thing is that you have all the CI checks passing, and for the content try to keep it short and Emerging talents program specific.

File: solutions/tests/README.md [hunk 4]
```diff
@@ -1 +1,146 @@
...
+   ensuring they perform as expected.
+
+### Naming Conventions
+
+- **Test files**: Should follow the pattern `test_<module_name>.py`.
+- **Test functions**: Should be named `test_<functionality>`, e.g., `test_login`.
+
+---
+
+## Testing Guidelines
```
Comment (Bikaze) [inline solutions/tests/README.md:36]: You can remove this dangling Heading

File: solutions/tests/README.md [hunk 5]
```diff
@@ -1 +1,146 @@
...
+
+- **Target Coverage**: Aim for high test coverage, but prioritize testing the
+   critical paths. Quality matters more than quantity—avoid excessive or
+  redundant tests.
+  
+- **Use Coverage Tools**: Regularly measure coverage using tools like
+   `coverage.py` or `pytest-cov` to ensure all edge cases, normal cases, and
+    error-handling scenarios are tested.
+
+### Test Dependencies
```
Comment (Bikaze) [inline solutions/tests/README.md:61]: Can you also delete the rest of this README because it is not particularly relevant to the project. Right?

File: solutions/tests/README.md [hunk 6]
```diff
@@ -1 +1,146 @@
...
+      pytest tests/functional/
+      ```
+
+    - End-to-end tests:
+
+      ```bash
+      pytest tests/e2e/
+      ```
+
+4. **Review Results**: Check the output to see which tests passed or failed.
```
Comment (Bikaze) [inline solutions/tests/README.md:145]: I think keeping this is good.

File: solutions/tests/README.md (hunk 5 above)
Comment (ziadahanass) [reply solutions/tests/README.md:61]: keep test coverage or remove it also?

Comment (Bikaze) [reply solutions/tests/README.md:61]: No, I think it makes sense and it is good to have it. Please who know it can use it.

Comment (Bikaze) [reply solutions/tests/README.md:61]: But you can decide whether to keep it or not.

Comment (ziadahanass) [reply solutions/tests/README.md:61]: i am gonna push the edited here, and then if anyone else have any opinion we can talk about

**Review by Bikaze — APPROVED** [review]
Wow this is perfect!
